enhancements/installer/storage-class-encrypted.md

```diff
@@ -0,0 +1,308 @@
+so that the default storage class will be set to provision volumes with those keys.
+
+### Goals
+
+* Default storage class is set to use user-provided keys at cluster creation for public clouds (AWS, GCP, Azure)
+
+### Non-Goals
+* Other clouds besides AWS, GCP, & Azure are out of scope
+
+
+## Proposal
+
+### Workflow Description
+
+1. Cluster creator specifies (aws|gcp|azure) user-managed encryption keys 
+in `defaultMachinePlatform` of install config.
+2. Installer creates a manifest for `ClusterCSIDriver` which includes encryption keys in spec.
+3. Cluster-storage-operator/CSI drivers use installer-created `ClusterCSIDriver`
+to create default storage class using user-encrypted keys.
+
+
+### API Extensions
+
+This enhancement extends the [`ClusterCSIDriver` API object](https://github.com/openshift/api/blob/efeef9d83325a0ba12a636e5efd7d9414036f9f7/operator/v1/types_csi_cluster_driver.go).
+
+AWS, Azure & GCP will be added as new valid values for the `CSIDriverType` union discriminator
+in the CSIDriverConfigSpec:
+
+```go
+// CSIDriverType indicates type of CSI driver being configured.
+// +kubebuilder:validation:Enum="";vSphere
+type CSIDriverType string
+
+const (
+  AWSDriverType     CSIDriverType = "aws"
+  AzureDriverType   CSIDriverType = "azure"
+  GCPDriverType     CSIDriverType = "gcp"
+  VSphereDriverType CSIDriverType = "vSphere" // existing platform
+)
+
+// CSIDriverConfigSpec defines configuration spec that can be
+// used to optionally configure a specific CSI Driver.
+// +union
+type CSIDriverConfigSpec struct {
+  // driverType indicates type of CSI driver for which the
+  // driverConfig is being applied to.
+  //
+  // Valid values are:
+  //
+  // * aws
+  // * azure
+  // * gcp
+  // * vSphere
+  //
+  // ---
+  // Consumers should treat unknown values as a NO-OP.
+  //
+  // +kubebuilder:validation:Required
+  // +unionDiscriminator
+  DriverType CSIDriverType `json:"driverType"`
+
+  // AWS is used to configure the AWS CSI driver.
+  // +optional
+  AWS *AWSCSIDriverConfigSpec `json:"aws,omitempty"`
+
+  // Azure is used to configure the Azure CSI driver.
+  // +optional
+  Azure *AzureCSIDriverConfigSpec `json:"azure,omitempty"`
+
+  // GCP is used to configure the GCP CSI driver.
+  // +optional
+  GCP *GCPCSIDriverConfigSpec `json:"gcp,omitempty"`
+
+  ... exisiting vsphere field ... 
+}
+```
+
+ and fields for the encryption keys will be added
+to platform-specific config specs:
+
+#### AWS
+
+```go
+// AWSCSIDriverConfigSpec defines properties that can be configured for the AWS CSI driver.
+type AWSCSIDriverConfigSpec struct {
+    // KMSKeyARN sets the cluster default storage class to encrypt volumes with a user-defined KMS key,
+    // rather than the default KMS key used by AWS.
+    // According to https://docs.aws.amazon.com/service-authorization/latest/reference/list_awskeymanagementservice.html
+    // the format for a KMS key ARN is:
+    // arn:${Partition}:kms:${Region}:${Account}:key/${KeyId}
```

Comment:
Can we create a `kubebuilder:validation:pattern` for this?

Reply:
During my last revision I considered this and briefly touched on it here: https://github.com/openshift/enhancements/pull/1163#discussion_r1067263018

I was unable to find existing regexes used for ARN validation, and considered trying to craft one. I am willing to try to write one up, but I do have some reservations:

1. In my personal experience, using static validation (i.e. does the input fit this regex) is an unnecessary failure point when dynamic validation (i.e. can the AWS API find the supplied ARN) covers all the necessary bases. My experience in this may not be relevant, though, when defining an API such as this.
2. I think some of the subcomponents, e.g. `Partition` may be difficult to pin down a meaningful pattern
3. My preliminary research didn't return any meaningful prior art in our API or in other codebases, which makes me suspect regex validation of ARNs is not common.

That said, if this would be beneficial I can do my best to craft one up, at least breaking each subcomponent down as much as possible.

Reply:
My counterpoints with my API review hat on
> In my personal experience, using static validation (i.e. does the input fit this regex) is an unnecessary failure point when dynamic validation (i.e. can the AWS API find the supplied ARN) covers all the necessary bases. My experience in this may not be relevant, though, when defining an API such as this.

With dynamic validation, a user can persist invalid data into the API, so they may make and edit, and go away and not bother to check. With static validation in the API like this, they cannot persist the data if it's in the wrong format, which gives them an immediate error (a webhook would achieve the same), so giving the feedback earlier is a UX win

> I think some of the subcomponents, e.g. Partition may be difficult to pin down a meaningful pattern

For something like this, we can either try and pin it down, or just include all characters we think appropriate. Having the validation for `arn:.*:kms:[a-z0-9]+(-[a-z0-9]+)*:[0-9]{12}:key/.*` would be a good start I would say.

That said, according to the docs, there are only three supported partitions currently, `aws`, `aws-cn`, `aws-us-gov`, https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html, so we can probably do better

> My preliminary research didn't return any meaningful prior art in our API or in other codebases, which makes me suspect regex validation of ARNs is not common.

We are making a conscious effort to improve validations in newer APIs, but yes, fair point. I think a number of other APIs break these things down. For example, does the installer already know which partition, region and account it's installing the cluster into? In which case you could construct the ARN from the data by just asking for the Key ID right? Could be a better UX.

Note each account is tied to a specific partition, so we could ask for region/account/key ID and then look up the partition using an AWS API (bit of an assumption there)

Reply:
All of this makes a lot sense.

> For example, does the installer already know which partition, region and account it's installing the cluster into? In which case you could construct the ARN from the data by just asking for the Key ID right? Could be a better UX.

Yes, but the installer is already asking for the full ARN, so I think the more important question is can the CSO/CSI Driver determine those values for the UX when users are adding this Day 2. @dobsonj and I talked about this in the context of whether to keep Azure fields like resource group optional; and we decided to keep those fields required, when they could be determined dynamically. My thought was that we could always step the fields down to optional, once we grow the capability to determine the defaults; but I see value in not even defining the field in the API and keeping it as minimal as possible.